engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java

```diff
@@ -179,16 +195,34 @@ static Object doUncached(
   }
 
   @ExportMessage
-  Text toDisplayString(boolean allowSideEffects, @CachedLibrary("this") InteropLibrary atoms) {
+  Text toDisplayString(
+    boolean allowSideEffects,
+    @CachedLibrary("this") InteropLibrary atoms,
+    @CachedLibrary(limit="3") WarningsLibrary warnings,
+    @Cached BranchProfile handleError
+  ) {
+    Object result = null;
+    String msg;
     try {
-      return TypesGen.expectText(atoms.invokeMember(this, "to_text"));
-    } catch (UnsupportedMessageException
-        | ArityException
-        | UnknownIdentifierException
-        | UnsupportedTypeException
-        | UnexpectedResultException e) {
-      return Text.create(this.toString(10));
+      result = atoms.invokeMember(this, "to_text");
+      if (warnings.hasWarnings(result)) {
+        result = warnings.removeWarnings(result);
+      }
+      if (result instanceof DataflowError) {
+        msg = this.toString("Error in method `to_text` of [", 10, "]: ", result);
+      } else {
+        return TypesGen.expectText(result);
+      }
+    } catch (AbstractTruffleException panic) {
+      handleError.enter();
+      msg = this.toString("Panic in method `to_text` of [", 10, "]: ", panic);
+    } catch (
+      UnexpectedResultException | UnsupportedMessageException | ArityException | UnknownIdentifierException | UnsupportedTypeException e
+    ) {
+      handleError.enter();
+      msg = this.toString("Error in method `to_text` of [", 10, "]: Expected Text but got ", result);
```

Comment:
> I cannot provide a decent test case for these other exceptions, I'm not sure how to trigger them in that context.

That's my problem as well. Triggering these exceptions isn't easy, but they have to be caught, as they are checked exception.

Reply:
Rewritten in [44978cd](https://github.com/enso-org/enso/pull/4034/commits/44978cd0c59705bc662a75a1fd14db4cd83c9398). Now it will print `Panic...` in case of those _strange exceptions_.